README
Copyright (C) 2008  Matteo Vescovi <matteo.vescovi@yahoo.co.uk>
___________________
The Presage project
~~~~~~~~~~~~~~~~~~~

Welcome to Presage, the intelligent predictive text entry system.


This document will guide you through the steps required to configure
and build the Presage system and libraries. You should be ready to run
Presage in a few minutes.

This README does not contain the answers to all the questions you
might have. If you run into troubles, if you have problems, or if you
would like a pressing question answered, please direct your attention
to the doc/ directory, where you will find all the available
documentation.


Requirements:
- A working development enviroment.
  On Linux/Unix systems, this means having GCC (g++), the GNU Compiler
  Collection packages, as well as libraries for these languages
  (libstdc++, glibc, etc).

  On Solaris systems, this means having Sun Studio 10 or Sun Studio 11
  compilers.

  On Windows systems, please either install the Cygwin development
  environment or the MinGW/MSYS developer environment.

  Cygwin can be obtained from http://www.cygwin.com
  Instructions on how to setup Cygwin and required packages can be
  found in the doc/INSTALL_Cygwin_dev_env.txt file.

  MinGW/MSYS can be obtained from http://www.mingw.org
  Instructions on how to setup MinGW/MSYS and required packages can be
  found in the doc/INSTALL_MinGW_MSYS_dev_env.txt file.

- SQLite package. SQLite is a small C library that implements a
  self-contained, embeddable, zero-configuration SQL database engine.
  SQLite can be obtained from http://www.sqlite.org
 
  On Debian systems, install libsqlite3-dev and sqlite3 packages.

Optional:
- Curses library. On Debian systems, install libncurses5-dev.

- CPPUnit library. On Debian systems, install libcppunit-dev.

- Python package. On Debian systems, install python-dev.

- SWIG package. On Debian systems, install swig.

- help2man package. On Debian systems, install help2man.

- doxygen package. On Debian systems, install doxygen.

- dot program. On Debian systems, install graphviz.

- X11 EvIE extension library. On Debian systems, install libxevie-dev.

- GNOME development package. On Debian systems, install gnome-devel.

- CMU-Cambridge Statistical Language Modeling toolkit.


_______________________
STEP 0: Getting Presage
~~~~~~~~~~~~~~~~~~~~~~~

Official releases and development snapshots are available from the
download section of the official website hosted by SourceForge

  http://presage.sourceforge.net

Alternatively, get the latest and greatest sources from SVN.


_________________________
STEP 1: Unpacking Presage
~~~~~~~~~~~~~~~~~~~~~~~~~

On Linux/Unix, uncompress and untar the distribution with 

  tar jxvf presage-*.tar.bz2

or

  tar zxvf presage-*.tar.gz

depending on whether you downloaded the tarball or the bzipped archive.


___________________________
STEP 2: Configuring Presage
~~~~~~~~~~~~~~~~~~~~~~~~~~~

Configuring Presage to different platforms, environments and
operating systems is simply a matter of running

  ./configure

in the top-level directory of the distribution. This will configure
Presage for your system. Should you require additional options, try

  ./configure --help

to see the available options.


_________________________
STEP 3: Compiling Presage
~~~~~~~~~~~~~~~~~~~~~~~~~

Kick off the compilation with

  make

This will build the Presage system, libraries and available
plugins.


__________________________
STEP 4: Installing Presage
~~~~~~~~~~~~~~~~~~~~~~~~~~

Install Presage and its support libraries and plugins by typing

  make install


__________________________________
STEP 5: Let Presage read your mind
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

Presage includes some demonstration programs that will read the user
input and guess what the user intends to type using its predictive
technology. These programs require additional libraries that are not
needed by the presage library and might not be configured to build,
depending on the build system installed libraries (except
presage_demo_text, which has no extra dependencies and is therefore
always built). To start, type:

 * presage_demo_text

   Simple textual console application designed to interactively test
   presage predictive functionality. The program reads strings from
   standard input and returns predictions.


 * presage_demo

   Ncurses based demo program similar to presage_demo_text.


 * presage_simulator

   Console application that simulates presage predictive process and
   reports on predictive performance.


 * presage_python_demo

   Simple console application designed to interactively test presage
   python binding.


 * gprompter

   Cross-platform predictive text editor. gprompter displays
   predictions in a contextual pop-up box as each letter is
   typed. Predictions can be easily selected and inserted in the
   document. gprompter is a cross-platform GTK+ application.


 * pyprompter

   Text editor with autocompletion popups displaying predictions and
   enabling users to quickly and efficiently select the correct
   prediction or requesting more predictions. pyprompter is a
   cross-platform wxPython application.


 * pypresagemate

   Universal predictive text companion. Pypresagemate works alongside
   any AT-SPI aware application. The Assistive Technology Service
   Provider Interface (AT-SPI) is a toolkit-neutral way of providing
   accessibility facilities in applications. Pypresagemate works in
   the background by tracking what keystrokes are typed and displaying
   predictions in its window. When a prediction is selected, text is
   sent to the active application.

 * gpresagemate

   X application that monitors all X events and display predictions in
   its window. When users select the correct prediction, it is sent to
   the current active X application as if it had been typed
   in. Requires that XEvIE X extension is enabled.


########/

Copyright (C) 2008  Matteo Vescovi <matteo.vescovi@yahoo.co.uk>

Presage is free software; you can redistribute it and/or modify
it under the terms of the GNU General Public License as published by
the Free Software Foundation; either version 2 of the License, or
(at your option) any later version.

This program is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU General Public License for more details.

You should have received a copy of the GNU General Public License along
with this program; if not, write to the Free Software Foundation, Inc.,
51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.

########\


NEWS
Copyright (C) 2014  Matteo Vescovi <matteo.vescovi@yahoo.co.uk>
___________________
The Presage project
~~~~~~~~~~~~~~~~~~~

NEWS
----


Presage 0.9.1 released
----------------------
Presage 0.9.1 is now available for download.

Presage 0.9.1 introduces a new offline learning API method. Active
predictors in presage are capable of online learning (dynamic learning
triggered by context changes - context-awareness enables presage to
train predictors on changes in context dynamically). This new offline
learning method provides a way to instruct presage to learn from a
specific body of text, separate from the context. A corresponding
ONLINE_LEARNING configuration variable has been added to control
whether online machine learning is enabled or disabled for
applications that require control over what is learnt. Moreover,
learning performance has been improved by a factor of 10 in most
scenarios.

Presage 0.9.1 provides a new Windows Communication Foundation
service. Applications can now consume presage's predictive
functionality by invoking on the WCF service process rather than
directly linking to presage library. The presage WCF service
application is automatically started at logon and displays a
notification icon in the Windows system tray. The notification icon
can be used to shutdown the presage WCF service or to display further
information. The presage WCF service can only be consumed by the same
user that started the service and that is currently logged on the
local system. The service uses the NetNamedPipeBinding
binding/transport, "a secure, reliable, optimized binding that is
suitable for on-machine communication between WCF applications".
The presage predictive service endpoint is exposed at this address:
net.pipe://localhost/PresageService/v1/presage
The presage Metadata Exchange endpoint is exposed at this address:
net.pipe://localhost/PresageService/v1/mex

Presage 0.9.1 adds support for in-memory SQLite databases. Users
wishing to store a temporary language model in memory can do so by
using the started in-memory database filename designator ":memory:".

Presage 0.9.1 brings several fixes and improvements to Windows
installer/uninstaller.

Presage 0.9.1 also includes other bug fixes and enhancements. Please
see the ChangeLog for further details.


Presage 0.9 released
--------------------
Presage 0.9 is now available for download.

Presage 0.9 features a new .NET binding. .NET applications can now
easily integrate with presage by using the presage.NET assembly. The
native C++ presage core is wrapped using P/Invoke functionality
provided by the .NET framework. A C# demo application illustrates how
simple it is to use presage.net in .NET (bindings/c#).

Presage 0.9 now supports the MinGW-w64 64-bit and 32-bit toolchains
[http://mingw-w64.sourceforge.net/] from the MinGW-builds
project. Detailed instructions on how to build with the MinGW-w64
toolchains are available in doc/INSTALL_MinGW64_MSYS_dev_env.txt and
doc/INSTALL_MinGW32_MSYS_dev_env.txt. Pre-built Windows installers are
also available.

Presage 0.9 fixes several issues uncovered by static code analysis
tools. Enhancements to the Notepad++ plugin, the Windows installer,
and the build system have been added.

Presage 0.9 also includes other bug fixes and enhancements. Please see
the ChangeLog for further details.


Presage 0.8.9 released
----------------------
Presage 0.8.9 is now available for download.

Presage 0.8.9 comes with a number of small enhancements and several
fixes. No major features were added in this release. Presage 0.8.9 is
a minor mantainance release.

Presage 0.8.9 fixes several issues uncovered by static code analysis
tools. Enhancements to the Notepad++ plugin and the Windows installer
have been added. Bugs affecting the concurrent execution of the
internal testsuite have been fixed. Issues affecting the build system
have been fixed.

Presage 0.8.9 also includes other bug fixes and enhancements. Please
see the ChangeLog for more details.


Presage 0.8.8 released
----------------------
Presage 0.8.8 is now available for download.

Presage 0.8.8 allows to enabled multiple instances of the same
predictor class at runtime through configuration. Each predictor
instance is independently configured to allow greater
flexibility. Predictive performance can be more finely tailor to the
users' specific needs by tuning the resources and configuration of
each predictor. The default configuration has been changed to add a
custom user smoothed ngram predictor, which adaptively learns its
language model. The user-specific default configuration file location
is now ~/.presage/presage.xml (~/.presage.xml is no longer used). The
configuration now supports and expands environment variables.

Presage 0.8.8 adds support for filters to dictionary predictor,
recency predictor, abbreviation expansion predictor, and dejavu
predictor. The smoothed n-gram predictor will create and initialize a
language model database if it does not already exist. Learning
performance of the smoothed n-gram predictor has been improved. A new
configuration variable has been added to control context tracking case
sensitivity.

Presage 0.8.8 fixes a bug that caused learning of spurious text. A bug
affecting gprompter when learning is enabled was fixed. A memory
allocation bug and a memory leak were fixed in the Notepad++ plugin.

Presage 0.8.8 ships with an updated embedded XML configuration parser
library and update text editing component.

Presage 0.8.8 comes with a number of other enhancements and
fixes. Please see the ChangeLog for more details.


Presage 0.8.7 released
----------------------
Presage 0.8.7 is now available for download.

Presage 0.8.7 now integrates into the Notepad++, a powerful Windows
text editor, thanks to the new presage predictive Notepad++ plugin,
NppPresage. NppPresage is able to detect a presage installation and
dynamically load presage to add predictive text functionality to
Notepad++.


src/lib/presage.h

/******************************************************
 *  Presage, an extensible predictive text entry system
 *  ---------------------------------------------------
 *
 *  Copyright (C) 2008  Matteo Vescovi <matteo.vescovi@yahoo.co.uk>

    This program is free software; you can redistribute it and/or modify
    it under the terms of the GNU General Public License as published by
    the Free Software Foundation; either version 2 of the License, or
    (at your option) any later version.

    This program is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
    GNU General Public License for more details.

    You should have received a copy of the GNU General Public License along
    with this program; if not, write to the Free Software Foundation, Inc.,
    51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.
                                                                             *
                                                                **********(*)*/


#ifndef PRESAGE
#define PRESAGE

#include "presageException.h"
#include "presageCallback.h"

/** \mainpage

    \section intro_section Introduction

    Presage is an intelligent predictive text entry
    platform. Presage exploits <a
    href="http://en.wikipedia.org/wiki/Redundancy_%28information_theory%29">redundant
    information</a> embedded in natural languages to generate
    predictions. Presage's modular and pluggable <a
    href="?q=node/14">architecture</a> allows its <a
    href="http://en.wikipedia.org/wiki/Language_modeling">language
    model</a> to be extended and customized to utilize statistical,
    syntactic, and semantic information sources.

    A predictive text entry system attempts to improve ease and speed of
    textual input. Word prediction consists in computing which word tokens
    or word completions are most likely to be entered next. The system
    analyses the text already entered and combines the information thus
    extracted with other information sources to calculate a set of most
    probable tokens.

    A typical presage-based application would display the set of most
    probable tokens (i.e. a list of suggestions) to the user and
    automatically enter the desired token after the user selects it. If
    the list of suggestions does not contain the desired word, the user
    continues entering text until the correct suggestion is offered or
    until the user is done entering text.

    Presage is fundamentally different from predictive <a
    href="http://en.wikipedia.org/wiki/T9_%28predictive_text%29">input
    technologies commonly found on mobile phones</a>, which might more
    accurately be described as 'disambiguating text entry' rather than
    'predictive text entry' systems.  Such systems do not try to guess
    what the user intends to write in the future, only to determine what
    they most-likely intend to write in the present, given their past
    input.  Presage, on the other hand, actively predicts the what the
    user intends to write, and only reverts to <a
    href="http://en.wikipedia.org/wiki/Word_completion">word
    completion</a> mode if the prediction did not contain the desired
    token.

    Presage is <a
    href="http://www.gnu.org/philosophy/free-sw.html">free
    software</a>. It is distributed under the term of the <a
    href="http://www.gnu.org/copyleft/gpl.html">General Public
    License</a>.

    \author Matteo Vescovi

    \section getting_started_section Getting started
    \include getting_started.txt

*/

/********************************
 *    Presage C++ API starts here
 */

#ifdef __cplusplus
#ifndef _MSC_VER

#include <string>
#include <vector>
#include <map>

/* Forward declarations, not part of presage C++ API */
class Configuration;
class ProfileManager;
class Profile;
class ContextTracker;
class PredictorRegistry;
class PredictorActivator;
class Selector;

/** \brief Presage, the intelligent predictive text entry platform.
 */
class Presage {
public:
    /** Creates and initializes presage.
     *
     * \param callback is a user-supplied implementation of PresageCallback interface
     * 
     * Presage does not take ownership of the callback object.
     */
    Presage(PresageCallback* callback) throw (PresageException);


    /** Creates and initializes presage with supplied configuration.
     *
     * \param callback is a user-supplied implementation of PresageCallback interface
     * \param config path to configuration file
     *
     * Presage does not take ownership of the callback object.
     */
    Presage(PresageCallback* callback, const std::string config) throw (PresageException);


    /** Destroys presage.
     */
    ~Presage();

    /** \brief Obtain a prediction.
     *
     * This method requests that presage generates a prediction based
     * on the current context.
     *
     * \return prediction (vector of strings) based on the current
     * context.
     *
     */
    std::vector<std::string> predict() throw (PresageException);

    /** \brief Obtain a prediction that matches the supplied token
     *         filter.
     *
     * \param filter a vector of strings to use to filter the
     * prediction for desired tokens.  I.e. If the current prefix is
     * "gr" and the filter is ["ea", "an"], then only words starting
     * with "grea" or "gran" such as "great" or "grand" will be
     * returned in the prediction.
     *
     * \return prediction containing only tokens that begin with one
     * of the filter tokens.
     *
     */
    std::multimap<double, std::string> predict(std::vector<std::string> filter) throw (PresageException);

    /** \brief Learn from text offline.
     *
     * Requests presage to offline learn from \param text. Active
     * predictors in presage are capable of online learning (dynamic
     * learning triggered by context changes - context-awareness
     * enables presage to train predictors on changes in context
     * dynamically). This method provides a way to instruct presage to
     * learn from a specific body of text, separate from the context.
     *
     * \param text a text string to learn from.
     *
     */
    void learn(const std::string text) const throw (PresageException);

    /** \brief Callback getter/setter.
     *
     * \param callback to be used by presage (pass a null pointer to
     * obtain callback to current callback without modifying it)
     *
     * \return pointer to previously used callback
     */
    PresageCallback* callback(PresageCallback* callback) throw (PresageException);

    /** \brief Request presage to return the completion string for the given predicted token.
     *
     * Requests presage to return the completion string. The
     * completion string is defined as the string which, when appended
     * to the current prefix, forms the token passed as the str
     * argument.
     *
     * \param str successful prediction, for which a completion string
     * is requested
     *
     * \return completion string
     */
    std::string completion(std::string str) throw (PresageException);

    /** \brief Returns the text entered so far.
     *
     * \return context, text entered so far.
     */
    std::string context() const throw (PresageException);

    /** \brief Returns true if a context change occured.
     *
     * \return true if a context change occured after the last update
     * or predict calls, or false otherwise.
     */
    bool context_change() const throw (PresageException);

    /** \brief Returns the current prefix.
     *
     * \return prefix
     */
    std::string prefix() const throw (PresageException);

    /** \brief Gets the value of specified configuration variable.
     *
     * Programmatically get the value currently assigned to the
     * specified configuration \param variable
     *
     * \return value assigned to configuration variable.
     */
    std::string config(const std::string variable) const throw (PresageException);

    /** \brief Sets the value of specified configuration variable.
     *
     * Programmatically set the specified configuration \param
     * variable to \param value . This will override the setting read
     * from the configuration file in use.
     *
     */
    void config(const std::string variable, const std::string value) const throw (PresageException);

    /** \brief Save current configuration to file.
     *
     * Call this method to persist current presage configuration to
     * file. The configuration data will be saved to the currently
     * active XML profile.
     *
     */
    void save_config() const throw (PresageException);

    /*
     * Presage public API ends here
     */

private:
    ProfileManager*     profileManager;
    Configuration*      configuration;
    PredictorRegistry*  predictorRegistry;
    ContextTracker*     contextTracker;
    PredictorActivator* predictorActivator;
    Selector*           selector;

};

#endif /* _MSC_VER */
#endif /* __cplusplus */

/*
 *    Presage C++ API ends here
 *******************************/


/*******************************
 *    Presage C API starts here
 */

#ifdef __cplusplus
extern "C" {
#endif

    typedef struct _presage* presage_t;

    presage_error_code_t presage_new                 (_presage_callback_get_past_stream past_stream_cb,
                                                      void* past_stream_cb_arg,
                                                      _presage_callback_get_future_stream future_stream_cb,
                                                      void* future_stream_cb_arg,
                                                      presage_t* result);
    
    presage_error_code_t presage_new_with_config     (_presage_callback_get_past_stream past,
                                                      void* past_stream_cb_arg,
                                                      _presage_callback_get_future_stream future_stream_cb,
                                                      void* future_stream_cb_arg,
                                                      const char* config,
                                                      presage_t* result);
    
    void                 presage_free                (presage_t prsg);
    
    void                 presage_free_string         (char* str);
    void                 presage_free_string_array   (char** str);

    presage_error_code_t presage_predict             (presage_t prsg,
                                                      char*** result);
    
    presage_error_code_t presage_learn               (presage_t prsg,
						      const char* text);

    presage_error_code_t presage_completion          (presage_t prsg,
                                                      const char* token,
                                                      char** result);

    presage_error_code_t presage_context             (presage_t prsg, 
                                                      char** result);

    presage_error_code_t presage_context_change      (presage_t prsg,
                                                      int* result);

    presage_error_code_t presage_prefix              (presage_t prsg,
                                                      char** result);

    presage_error_code_t presage_config              (presage_t prsg,
                                                      const char* variable,
                                                      char** result);

    presage_error_code_t presage_config_set          (presage_t prsg,
                                                      const char* variable,
                                                      const char* value);
    
    presage_error_code_t presage_save_config         (presage_t prsg);

#ifdef __cplusplus
}
#endif

/*
 *    Presage C API ends here
 ****************************/


#endif /* PRESAGE */
